File: C++/CloneGraph.cc
// Clone an undirected graph. Each node in the graph contains a label and a list of its neighbors.

 struct UndirectedGraphNode {
     int label;
     vector<UndirectedGraphNode *> neighbors;
     UndirectedGraphNode(int x) : label(x) {};
 };

class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (!node) {
            return nullptr;
        }
        queue<UndirectedGraphNode*> q;
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> node_map;

        q.push(node);
        UndirectedGraphNode* new_root = new UndirectedGraphNode(node->label);
        node_map.emplace(node, new_root);

        while (!q.empty()) {
            UndirectedGraphNode* old_node = q.front();
            UndirectedGraphNode* new_node = node_map[old_node];
            q.pop();
            for (auto n : old_node->neighbors) {
                if (node_map.find(n) == node_map.end()) {
                    UndirectedGraphNode* u = new UndirectedGraphNode(n->label);
                    q.push(n);
                    node_map.emplace(n, u);
                }
                new_node->neighbors.push_back(node_map[n]);
            }
        }
        return new_root;
    }
};

File: Java/CloneGraph.java
// Clone an undirected graph. Each node in the graph contains a label and a list of its neighbors.

public class CloneGraph {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) {
            return null;
        }
        Map<UndirectedGraphNode, UndirectedGraphNode> nodeMap = new HashMap<>();
        Queue<UndirectedGraphNode> q = new LinkedList<>();
        q.add(node);
        UndirectedGraphNode clone = new UndirectedGraphNode(node.label);
        nodeMap.put(node, clone);
        while (!q.isEmpty()) {
            UndirectedGraphNode u = q.poll();
            for (UndirectedGraphNode v : u.neighbors) {
                if (!nodeMap.containsKey(v)) {
                    UndirectedGraphNode w = new UndirectedGraphNode(v.label);
                    nodeMap.put(v, w);
                    q.add(v);
                }
                nodeMap.get(u).neighbors.add(nodeMap.get(v));
            }
        }
        return clone;
    }
}
